cpu/qn908x/periph/rtc.c

```diff
@@ -0,0 +1,175 @@
+        DEBUG("rtc_init() ppm=%" PRIi32 "\n", (int32_t)ppm);
+        uint32_t abs_ppm = abs(ppm);
+        if (abs_ppm >= (1u << 16)) {
+            abs_ppm = (1u << 16) - 1;
+        }
+
+        RTC->CAL = RTC_CAL_PPM(abs_ppm) | (ppm >= 0 ? RTC_CAL_DIR_MASK : 0);
+        RTC->CTRL |= RTC_CTRL_CAL_EN_MASK;
+    }
+
+    /* RTC clock (BIV) starts enabled after reset anyway. */
+    CLOCK_EnableClock(kCLOCK_Biv);
+
+    RTC->CTRL &= ~RTC_CTRL_SEC_INT_EN_MASK;
+
+    /* We only use the RTC_SEC_IRQ which triggers every second if enabled by the
+     * alarm. */
+    NVIC_EnableIRQ(RTC_SEC_IRQn);
+}
+
+int rtc_set_time(struct tm *time)
+{
+    uint32_t ts = rtc_mktime(time);
+
+    DEBUG("rtc_set_time(%" PRIu32 ")\n", ts);
+    /* Writing 1 to CFG in CTRL register sets the CNT 0 timer to 0, resetting
+     * the fractional part of the second, meaning that "SEC" is a round number
+     * of second when this instruction executes. */
+    RTC->CTRL |= RTC_CTRL_CFG_MASK;
+    RTC->SEC = ts;
+    while (RTC->STATUS &
+           (RTC_STATUS_SEC_SYNC_MASK | RTC_STATUS_CTRL_SYNC_MASK)) {}
+    return 0;
+}
+
+int rtc_get_time(struct tm *time)
+{
+    uint32_t ts = RTC->SEC;
+
+    DEBUG("rtc_get_time() -> %" PRIu32 "\n", ts);
+    rtc_localtime(ts, time);
+    return 0;
+}
+
+int rtc_set_alarm(struct tm *time, rtc_alarm_cb_t cb, void *arg)
+{
+    uint32_t ts = rtc_mktime(time);
+
+    DEBUG("rtc_set_alarm(%" PRIu32 ", %p, %p)\n", ts, cb, arg);
+
+    rtc_ctx.cb = cb;
+    rtc_ctx.arg = arg;
+    rtc_ctx.alarm = ts;
+
+    RTC->CTRL |= RTC_CTRL_SEC_INT_EN_MASK;
+    /* Wait until the CTRL_SEC is synced. */
+    while (RTC->STATUS & RTC_STATUS_CTRL_SYNC_MASK) {}
+
+    return 0;
+}
+
+int rtc_get_alarm(struct tm *time)
+{
+    DEBUG("rtc_clear_alarm() -> %" PRIu32 "\n", rtc_ctx.alarm);
+    rtc_localtime(rtc_ctx.alarm, time);
+    return 0;
+}
+
+void rtc_clear_alarm(void)
+{
+    DEBUG("rtc_clear_alarm()\n");
+    /* Disable the alarm flag before clearing out the callback. */
+    RTC->CTRL &= ~RTC_CTRL_SEC_INT_EN_MASK;
+    rtc_ctx.cb = NULL;
+    rtc_ctx.alarm = 0;
+}
+
+void rtc_poweron(void)
+{
+    /* We don't power on/off the RTC since it is shared with the timer module.
+     * Besides the CLK_32K clock source for every peripheral that might use it,
+     * there isn't much to turn off here. */
+}
+
+void rtc_poweroff(void)
+{
+    /* TODO: Coordinate with the RTT module to turn off the RTC clock when
+     * neither one is in use. */
+}
+
+/**
+ * @brief Interrupt service declared in vectors_qn908x.h
+ *
+ * We can only generate an interrupt every second, so we check the alarm value
+ * every time. For a hardware based comparison use the timer instead.
+ */
+void isr_rtc_sec(void)
+{
+    if (RTC_STATUS_SEC_INT_MASK & RTC->STATUS) {
+        DEBUG("isr_rtc_sec at %" PRIu32 "\n", RTC->SEC);
+        /* Write 1 to clear the STATUS flag. */
+        RTC->STATUS |= RTC_STATUS_SEC_INT_MASK;
```

Comment:
as all the other bits are read-only, I guess you could save a cycle by doing

```suggestion
        RTC->STATUS = RTC_STATUS_SEC_INT_MASK;
```

Reply:
Some are "reserved", but given the "write one to clear" nature of the STATUS register it might be safer to do `=`.